ai/ai-react-app/src/components/Specific/ChatMessage.tsx

```diff
@@ -23,38 +37,176 @@ const getMessageText = (message: Content): string => {
+        title={`Sources: ${tooltipText}`}
+        data-source-indices={seg.chunkIndices.join(",")}
+      >
+        {currentSegmentText}
+        <sup className={styles.sourceSuperscript}>
+          [{seg.chunkIndices.join(",")}]
+        </sup>
+      </span>,
+    );
+    lastIndexProcessed = Math.max(lastIndexProcessed, seg.endIndex);
+  });
+
+  // Add any remaining un-highlighted text
+  if (lastIndexProcessed < text.length) {
+    outputNodes.push(text.substring(lastIndexProcessed));
+  }
+
+  return outputNodes;
+};
+
 /**
  * Renders a single chat message bubble, styled based on the message role ('user' or 'model').
  * It only renders messages that should be visible in the log (user messages, or model messages
  * containing text). Function role messages or model messages consisting only of function calls
  * are typically not rendered directly as chat bubbles.
  */
-const ChatMessage: React.FC<ChatMessageProps> = ({ message }) => {
+const ChatMessage: React.FC<ChatMessageProps> = ({
+  message,
+  groundingMetadata,
+}) => {
   const text = getMessageText(message);
   const isUser = message.role === "user";
   const isModel = message.role === "model";
 
-  // We render:
-  // 1. User messages (even if they only contain images/files, the 'user' role indicates an entry).
-  // 2. Model messages *only if* they contain actual text content.
-  // We *don't* render:
-  // 1. 'function' role messages (these represent execution results, not direct chat).
-  // 2. 'model' role messages that *only* contain function calls (these are instructions, not display text).
-  // 3. 'system' role messages (handled separately, not usually in chat history display).
-  const shouldRender = isUser || (isModel && text.trim() !== "");
+  const shouldRender =
+    isUser ||
+    (isModel && text.trim() !== "") ||
+    (isModel &&
+      groundingMetadata &&
+      (groundingMetadata.groundingChunks?.length || 0 > 0));
 
   if (!shouldRender) {
     return null;
   }
 
+  let messageContentNodes: React.ReactNode[];
+  if (
+    isModel &&
+    groundingMetadata?.groundingSupports &&
+    groundingMetadata?.groundingChunks
+  ) {
+    messageContentNodes = renderTextWithInlineHighlighting(
+      text,
+      groundingMetadata.groundingSupports,
+      groundingMetadata.groundingChunks,
+    );
+  } else {
+    messageContentNodes = [text];
+  }
+
   return (
     <div
       className={`${styles.messageContainer} ${isUser ? styles.user : styles.model}`}
     >
       <div className={styles.messageBubble}>
-        {/* Use <pre> to preserve whitespace and newlines within the text content.
-                     Handles potential multi-line responses correctly. */}
-        <pre className={styles.messageText}>{text}</pre>
+        <pre className={styles.messageText}>
+          {messageContentNodes.map((node, index) => (
+            <React.Fragment key={index}>{node}</React.Fragment>
+          ))}
+        </pre>
+        {/* Source list rendering */}
```

Comment:
Do you think it's worth to put in a brief comment saying this is required by the terms of service or something? Could link to the devsite doc page if it's easy (if the timing isn't too tricky) but it's probably not vital.